src/JuliusSweetland.OptiKey.Core/Services/KeyboardOutputService.cs

```diff
@@ -490,12 +491,62 @@ public void ProcessSingleKeyText(string capturedText)
             }
         }
 
-        public void ProcessSingleKeyPress(string key, KeyPressKeyValue.KeyPressType type, int delayMs = 0)
+        public async Task ProcessSingleKeyPress(string inKey, KeyPressKeyValue.KeyPressType type)
         {
-            // TODO: This is a stub for future use
-            throw new NotImplementedException();            
+            Log.InfoFormat("Processing input value {0} with key press type {1}", inKey, type);
+
+            FunctionKeys funKey;
+            var virtualKeyCode = (Enum.TryParse(inKey, out funKey)) ? funKey.ToVirtualKeyCode() : null;
+
+            if (virtualKeyCode != null)
+            {
+                if (type == KeyPressKeyValue.KeyPressType.Press)
+                    publishService.KeyDown(virtualKeyCode.Value);
+                else if (type == KeyPressKeyValue.KeyPressType.Release)
+                    publishService.KeyUp(virtualKeyCode.Value);
```

Comment:
I eliminated KeyPressType.Toggle and made it so a Toggle action checks the KeyStateService and sends either Press or Release to the KeyboardOutputService.